File: exercises/domain-1/block-1.3/solution/solution.ts
type HookFn = (stepName: string, context: Record<string, unknown>) => boolean | void | Promise<boolean | void>

interface Step {
  name: string
  fn: (context: Record<string, unknown>) => Promise<Record<string, unknown>>
}

class WorkflowEngine {
  private steps: Step[] = []
  private preHooks: HookFn[] = []
  private postHooks: HookFn[] = []

  addStep(name: string, fn: Step["fn"]): this {
    this.steps.push({ name, fn })
    return this
  }

  addPreHook(hook: HookFn): this {
    this.preHooks.push(hook)
    return this
  }

  addPostHook(hook: HookFn): this {
    this.postHooks.push(hook)
    return this
  }

  async run(initialContext: Record<string, unknown> = {}): Promise<Record<string, unknown>> {
    let context = { ...initialContext }

    for (const step of this.steps) {
      // Run pre-hooks — if any return false, skip
      let skip = false
      for (const hook of this.preHooks) {
        const result = await hook(step.name, context)
        if (result === false) { skip = true; break }
      }

      if (!skip) {
        console.log(`▶ Running: ${step.name}`)
        context = await step.fn(context)
      } else {
        console.log(`⏭ Skipped: ${step.name}`)
      }

      // Run post-hooks
      for (const hook of this.postHooks) {
        await hook(step.name, context)
      }
    }

    return context
  }
}

const engine = new WorkflowEngine()

engine
  .addStep("read", async (ctx) => ({ ...ctx, document: "Hi", wordCount: 1 }))
  .addStep("transform", async (ctx) => ({ ...ctx, document: (ctx.document as string).toUpperCase() }))
  .addStep("save", async (ctx) => ({ ...ctx, saved: true }))

// Pre-hook: block save if wordCount < 5
engine.addPreHook((stepName, ctx) => {
  if (stepName === "save" && (ctx.wordCount as number) < 5) {
    console.log("  🚫 Pre-hook blocked save: document too short")
    return false
  }
})

// Post-hook: log completion
engine.addPostHook((stepName, ctx) => {
  console.log(`  ✓ Post-hook: ${stepName} done. Context keys: ${Object.keys(ctx).join(", ")}`)
})

engine.run({ filename: "test.txt" }).then((result) => {
  console.log("Final context:", result)
})

File: exercises/domain-1/block-1.3/solution/solution.py
from typing import Callable, Any
import asyncio

HookFn = Callable[[str, dict], bool | None]
StepFn = Callable[[dict], dict]


class WorkflowEngine:
    def __init__(self):
        self.steps: list[tuple[str, StepFn]] = []
        self.pre_hooks: list[HookFn] = []
        self.post_hooks: list[HookFn] = []

    def add_step(self, name: str, fn: StepFn) -> "WorkflowEngine":
        self.steps.append((name, fn))
        return self

    def add_pre_hook(self, hook: HookFn) -> "WorkflowEngine":
        self.pre_hooks.append(hook)
        return self

    def add_post_hook(self, hook: HookFn) -> "WorkflowEngine":
        self.post_hooks.append(hook)
        return self

    async def run(self, initial_context: dict = {}) -> dict:
        context = dict(initial_context)

        for name, step_fn in self.steps:
            skip = False
            for hook in self.pre_hooks:
                result = hook(name, context)
                if result is False:
                    skip = True
                    break

            if not skip:
                print(f"▶ Running: {name}")
                context = step_fn(context)
            else:
                print(f"⏭ Skipped: {name}")

            for hook in self.post_hooks:
                hook(name, context)

        return context


engine = WorkflowEngine()
engine.add_step("read", lambda ctx: {**ctx, "document": "Hi", "word_count": 1})
engine.add_step("transform", lambda ctx: {**ctx, "document": ctx["document"].upper()})
engine.add_step("save", lambda ctx: {**ctx, "saved": True})


def block_short_docs(step_name: str, ctx: dict) -> bool | None:
    if step_name == "save" and ctx.get("word_count", 0) < 5:
        print("  🚫 Pre-hook blocked save: document too short")
        return False


def log_completion(step_name: str, ctx: dict) -> None:
    print(f"  ✓ Post-hook: {step_name} done. Keys: {list(ctx.keys())}")


engine.add_pre_hook(block_short_docs)
engine.add_post_hook(log_completion)


async def main():
    result = await engine.run({"filename": "test.txt"})
    print("Final context:", result)


asyncio.run(main())
